pkg/queue/breaker.go

```diff
@@ -177,131 +176,136 @@ func (b *Breaker) Capacity() int {
 	return b.sem.Capacity()
 }
 
-// newSemaphore creates a semaphore with the desired maximal and initial capacity.
-// Maximal capacity is the size of the buffered channel, it defines maximum number of tokens
-// in the rotation. Attempting to add more capacity then the max will result in error.
-// Initial capacity is the initial number of free tokens.
+// newSemaphore creates a semaphore with the desired initial capacity.
 func newSemaphore(maxCapacity, initialCapacity int) *semaphore {
 	queue := make(chan struct{}, maxCapacity)
 	sem := &semaphore{queue: queue}
-	if initialCapacity > 0 {
-		sem.updateCapacity(initialCapacity)
-	}
+	sem.updateCapacity(initialCapacity)
 	return sem
 }
 
-// semaphore is an implementation of a semaphore based on Go channels.
-// The presence of elements in the `queue` buffered channel correspond to available tokens.
-// Hence the max number of tokens to hand out equals to the size of the channel.
-// `capacity` defines the current number of tokens in the rotation.
+// semaphore is an implementation of a semaphore based on packed integers and a channel.
+// state is an int64 that has two int32s packed into it: capacity and inFlight. The
+// former specifies how many request are allowed at any given time into the semaphore
+// while the latter refers to the currently in-flight requests.
+// Packing them both into one int64 allows us to optimize access semantics using atomic
+// operations, which can't be guaranteed on 2 individual values.
+// The channel is merely used as a vehicle to be able to "wake up" individual goroutines
+// if capacity becomes free. It's not consistently used in accordance to actual capacity
+// but is rather a communication vehicle to ensure waiting routines are properly worken
+// up.
 type semaphore struct {
-	queue    chan struct{}
-	reducers int
-	capacity int
-	mux      sync.RWMutex
+	state atomic.Int64
+	queue chan struct{}
 }
 
-// tryAcquire receives the token from the semaphore if there's one
-// otherwise an error is returned.
+// tryAcquire receives the token from the semaphore if there's one otherwise returns false.
 func (s *semaphore) tryAcquire() bool {
-	select {
-	case <-s.queue:
-		return true
-	default:
-		return false
+	for {
+		old := s.state.Load()
+		capacity, in := unpack(old)
+		if in >= capacity {
+			return false
+		}
+		in++
+		if s.state.CAS(old, pack(capacity, in)) {
+			return true
+		}
 	}
 }
 
-// acquire receives the token from the semaphore, potentially blocking.
+// acquire acquires capacity from the semaphore.
 func (s *semaphore) acquire(ctx context.Context) error {
-	select {
-	case <-s.queue:
-		return nil
-	case <-ctx.Done():
-		return ctx.Err()
+	for {
+		old := s.state.Load()
+		capacity, in := unpack(old)
+
+		if in >= capacity {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-s.queue:
```

Comment:
Im trying to figure out how we ensure here that we don't accidentally consume all the wake-ups. If we get woken up but then end up retrying on the CAS (just because something racily wrote to it) is that safe?

Reply:
Wouldn't just one of the pending ones wake up?

Reply:
That‘s precisely the point i‘m worrying about too 😅. I‘d love to either proof this to be wrong or false.

Reply:
My thinking was that the raciness would only come from releases and thus eventually the requests can actually pass through eventually. I‘d like it to err on the cautious side, i.e. when in doubt produce more wakeups than needed, the aquire code would handle that.

This might need an actual linearization analysis. Good old uni days 😂

Reply:
So I thought about this a little bit deeper:

You'll only run into waiting (and thus consuming a wakeup) if the state needs you to. So if you've already consumed one wakeup and end up having wait again, it means either that the capacity was lowered in between or that another request raced you to the acquire.

As such, I don't think this would consume more wakeups as are absolutely needed given that that condition cannot be reached by code that isn't required to wait by the current state.